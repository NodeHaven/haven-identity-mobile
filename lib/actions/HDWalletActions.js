// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { STORE_ROOT_ADDRESS, INC_IDENTITY_INDEX, INC_ACCOUNT_INDEX, SEED_CONFIRMED, ACCOUNT_RISK_SENT, RESET_HD_WALLET, STORE_ALL_KEYCHAIN_ADDRESS } from 'uPortMobile/lib/constants/HDWalletActionTypes'

// These are used to keep a track of used and unused paths for a HD wallet

// ADDING NEW ACTIONS PLEASE READ THIS!!!!
//
// If the action should be backed up add the `, _backup` attribute to the action.
// When should it be backed up?
//
// - is it backed by the reducer and not a saga?
// - is it something that can not be refreshed from the block chain etc? (eg. no need to backup nonces and balances)
// - is it not just related to UX? Some reducers are temporary and would not make sense on multiple device

export function storeRootAddress (root) {
  return {
    type: STORE_ROOT_ADDRESS,
    root,
    _backup: true
  }
}

export function incIdentityIndex () {
  return {
    type: INC_IDENTITY_INDEX,
    _backup: true
  }
}

export function incAccountIndex (identity) {
  return {
    type: INC_ACCOUNT_INDEX,
    identity,
    _backup: true
  }
}

export function seedConfirmed () {
  return {
    type: SEED_CONFIRMED,
    _backup: true
  }
}

export function accountRiskSent () {
  return {
    type: ACCOUNT_RISK_SENT,
    _backup: true
  }
}

export function resetHDWallet () {
  return {
    type: RESET_HD_WALLET
  }
}


export function storeAllKeyChainAddresses (addresses) {
  return {
    type: STORE_ALL_KEYCHAIN_ADDRESS,
    addresses
  }
}