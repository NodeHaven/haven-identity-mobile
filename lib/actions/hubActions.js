// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { STORE_HUB_HEAD, CATCHUP_HUB, QUEUE_EVENT, TAKE_QUEUED_EVENT, CLEAR_QUEUE, IMPORT_SNAPSHOT, RESET_HUB } from '../constants/HubActionTypes'

export function storeHubHead (hash) {
  return {
    type: STORE_HUB_HEAD,
    hash
  }
}
export function resetHub () {
  return {
    type: RESET_HUB
  }
}

export function catchupHub () {
  return {
    type: CATCHUP_HUB
  }
}

export function queueEvent (event) {
  return {
    type: QUEUE_EVENT,
    event
  }
}

export function takeQueuedEvent () {
  return {
    type: TAKE_QUEUED_EVENT
  }
}

export function clearQueue () {
  return {
    type: CLEAR_QUEUE
  }
}

export function importSnapshot (snapshot) {
  return {
    type: IMPORT_SNAPSHOT,
    snapshot
  }
}
