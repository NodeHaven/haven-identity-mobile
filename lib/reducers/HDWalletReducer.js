// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//

import { STORE_ROOT_ADDRESS, INC_IDENTITY_INDEX, INC_ACCOUNT_INDEX, SEED_CONFIRMED, ACCOUNT_RISK_SENT, RESET_HD_WALLET, STORE_ALL_KEYCHAIN_ADDRESS } from 'uPortMobile/lib/constants/HDWalletActionTypes'
import { IMPORT_SNAPSHOT } from 'uPortMobile/lib/constants/HubActionTypes'
import { RESET_DEVICE } from 'uPortMobile/lib/constants/GlobalActionTypes'
import { toClj, inc, assoc, updateIn, conj } from 'mori'

// This maintains information about the HD wallet needed to create new identities and accounts.
//
// Creates a hashMap similar to this:
// {
//   root: '0xabcde',
//   identities: [1, 3, 4]
// }
//
// Each element of the vector represents the path at it's index:
// The above example includes 3 identities each at position 0, 1 and 2.
//
// identity 0 has the most recent HD account path as 1. eg `m/7696500/0'/1'/0'`
// identity 1 has the most recent HD account path as 3. eg `m/7696500/1'/3'/0'`
// identity 2 has the most recent HD account path as 4. eg `m/7696500/2'/4'/0'`

// The path is as follows (the apostrophe denotes hardened derivation):

// m/version'/id'/acct'/purpose'

// * We set the `version` field to 7696500 to start with. This version field will allow us to update the path completely if we need to in the future.
// * The field `id` denotes the high-level identity. Starts at 0. Will have only one identity to start with, but this one can be incremented.
// * The field `acct` denotes the account of the identity. Account 0 will be used for the main off-chain identity. Ideally this account will not be used for any Ethereum related activities. Accounts 1 and above are for Ethereum related activities like holding ETH and tokens etc.
// * The field purpose denotes the purpose of the key. We use the following numbers:
//    * 0 - signing key (device key)
//    * 1 - recovery key
//    * 2 - asymmetric encryption key (may be defined only for the off-chain identity at first)

function HDWalletReducer (state = toClj({}), action) {
  switch (action.type) {
    case RESET_DEVICE:
      return toClj({})
    case STORE_ROOT_ADDRESS:
      return toClj({root: action.root, identities: [0]})
    case INC_IDENTITY_INDEX:
      return updateIn(state, ['identities'], (identities) => conj(identities, 0))
    case INC_ACCOUNT_INDEX:
      return updateIn(state, ['identities', action.identity], inc)
    case SEED_CONFIRMED:
      return assoc(state, 'seedConfirmed', true)
    case ACCOUNT_RISK_SENT:
      return assoc(state, 'accountRiskSent', true)
    case RESET_HD_WALLET: // used if something went wrong during onboarding or recovery
      return toClj({})
    case IMPORT_SNAPSHOT:
      if (action.snapshot.hdwallet && Object.keys(action.snapshot.hdwallet).length > 0) {
        return toClj(action.snapshot.hdwallet)
      } else {
        return state
      }
    case STORE_ALL_KEYCHAIN_ADDRESS:
      return assoc(state, 'seedAddresses', action.addresses)
    default:
      return state
  }
}

export default HDWalletReducer
