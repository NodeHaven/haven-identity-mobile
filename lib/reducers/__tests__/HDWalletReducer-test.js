// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import HDWalletReducer from 'uPortMobile/lib/reducers/HDWalletReducer.js'
import { storeRootAddress, incIdentityIndex, incAccountIndex, seedConfirmed, accountRiskSent, resetHDWallet } from 'uPortMobile/lib/actions/HDWalletActions.js'
import { toClj, toJs } from 'mori'
import { importSnapshot } from 'uPortMobile/lib/actions/hubActions.js'

const initialState = {
}

const populatedState = {
  root: '0x1234',
  identities: [3, 2, 1]
}

const reducer = (state, action) => toJs(HDWalletReducer(toClj(state), action))

it('performs STORE_ROOT_ADDRESS', () => {
  expect(reducer(undefined, storeRootAddress('0x1234'))).toMatchSnapshot()
  expect(reducer(initialState, storeRootAddress('0x1234'))).toMatchSnapshot()
  expect(reducer(populatedState, storeRootAddress('0x1234'))).toMatchSnapshot()
  expect(reducer(populatedState, storeRootAddress('0xabcde'))).toMatchSnapshot()
})

it('performs INC_IDENTITY_INDEX', () => {
  expect(reducer(undefined, incIdentityIndex())).toMatchSnapshot()
  expect(reducer(initialState, incIdentityIndex())).toMatchSnapshot()
  expect(reducer(populatedState, incIdentityIndex())).toMatchSnapshot()
  expect(reducer(populatedState, incIdentityIndex())).toMatchSnapshot()
})

it('performs INC_ACCOUNT_INDEX', () => {
  expect(reducer(populatedState, incAccountIndex(0))).toMatchSnapshot()
  expect(reducer(populatedState, incAccountIndex(1))).toMatchSnapshot()
  expect(reducer(populatedState, incAccountIndex(2))).toMatchSnapshot()
})

it('performs SEED_CONFIRMED', () => {
  expect(reducer(populatedState, seedConfirmed())).toMatchSnapshot()
})

it('performs ACCOUNT_RISK_SENT', () => {
  expect(reducer(populatedState, accountRiskSent())).toMatchSnapshot()
})

it('performs RESET_HD_WALLET', () => {
  expect(reducer(undefined, resetHDWallet())).toMatchSnapshot()
  expect(reducer(initialState, resetHDWallet())).toMatchSnapshot()
  expect(reducer(populatedState, resetHDWallet())).toMatchSnapshot()
})

it('performs IMPORT_SNAPSHOT', () => {
  expect(reducer(undefined, importSnapshot({hdwallet: toJs(populatedState)}))).toEqual(populatedState)
  expect(reducer(initialState, importSnapshot({hdwallet: toJs(populatedState)}))).toEqual(populatedState)
  expect(reducer(populatedState, importSnapshot({hdwallet: toJs(populatedState)}))).toEqual(populatedState)
  expect(reducer(populatedState, importSnapshot({hdwallet: {}}))).toEqual(populatedState)
})

it('ignores unsupported action', () => {
  expect(reducer(undefined, {type: 'UNSUPPORTED'})).toEqual(initialState)
  expect(reducer(initialState, {type: 'UNSUPPORTED'})).toEqual(initialState)
  expect(reducer(populatedState, {type: 'UNSUPPORTED'})).toEqual(populatedState)
})
