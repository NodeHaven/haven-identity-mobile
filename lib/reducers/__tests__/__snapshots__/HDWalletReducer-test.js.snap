// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`performs ACCOUNT_RISK_SENT 1`] = `
Object {
  "accountRiskSent": true,
  "identities": Array [
    3,
    2,
    1,
  ],
  "root": "0x1234",
}
`;

exports[`performs INC_ACCOUNT_INDEX 1`] = `
Object {
  "identities": Array [
    4,
    2,
    1,
  ],
  "root": "0x1234",
}
`;

exports[`performs INC_ACCOUNT_INDEX 2`] = `
Object {
  "identities": Array [
    3,
    3,
    1,
  ],
  "root": "0x1234",
}
`;

exports[`performs INC_ACCOUNT_INDEX 3`] = `
Object {
  "identities": Array [
    3,
    2,
    2,
  ],
  "root": "0x1234",
}
`;

exports[`performs INC_IDENTITY_INDEX 1`] = `
Object {
  "identities": Array [
    0,
  ],
}
`;

exports[`performs INC_IDENTITY_INDEX 2`] = `
Object {
  "identities": Array [
    0,
  ],
}
`;

exports[`performs INC_IDENTITY_INDEX 3`] = `
Object {
  "identities": Array [
    3,
    2,
    1,
    0,
  ],
  "root": "0x1234",
}
`;

exports[`performs INC_IDENTITY_INDEX 4`] = `
Object {
  "identities": Array [
    3,
    2,
    1,
    0,
  ],
  "root": "0x1234",
}
`;

exports[`performs RESET_HD_WALLET 1`] = `Object {}`;

exports[`performs RESET_HD_WALLET 2`] = `Object {}`;

exports[`performs RESET_HD_WALLET 3`] = `Object {}`;

exports[`performs SEED_CONFIRMED 1`] = `
Object {
  "identities": Array [
    3,
    2,
    1,
  ],
  "root": "0x1234",
  "seedConfirmed": true,
}
`;

exports[`performs STORE_ROOT_ADDRESS 1`] = `
Object {
  "identities": Array [
    0,
  ],
  "root": "0x1234",
}
`;

exports[`performs STORE_ROOT_ADDRESS 2`] = `
Object {
  "identities": Array [
    0,
  ],
  "root": "0x1234",
}
`;

exports[`performs STORE_ROOT_ADDRESS 3`] = `
Object {
  "identities": Array [
    0,
  ],
  "root": "0x1234",
}
`;

exports[`performs STORE_ROOT_ADDRESS 4`] = `
Object {
  "identities": Array [
    0,
  ],
  "root": "0xabcde",
}
`;
